bpf/lib/maps.h

```diff
@@ -186,21 +186,23 @@ struct bpf_elf_map __section_maps ENCRYPT_MAP = {
 	.max_elem	= 1,
 };
 
-struct egress_key {
-	struct bpf_lpm_trie_key lpm_key;
-	__u32 sip;
-	__u32 dip;
-};
-
 #ifdef ENABLE_EGRESS_GATEWAY
-struct bpf_elf_map __section_maps EGRESS_MAP = {
+struct bpf_elf_map __section_maps EGRESS_POLICY_MAP = {
 	.type		= LPM_MAP_TYPE,
-	.size_key	= sizeof(struct egress_key),
-	.size_value	= sizeof(struct egress_info),
+	.size_key	= sizeof(struct egress_policy_key),
+	.size_value	= sizeof(struct egress_policy),
 	.pinning	= PIN_GLOBAL_NS,
-	.max_elem	= EGRESS_MAP_SIZE,
+	.max_elem	= EGRESS_POLICY_MAP_SIZE,
 	.flags		= BPF_F_NO_PREALLOC,
 };
+
+struct bpf_elf_map __section_maps EGRESS_CT_MAP = {
```

Comment:
> Was there a discussion already about introducing a brand new CT map?

Yes, it was discussed in the initial brainstorming sessions I had with Daniel and Christopher, as well as in the CFP.

The reason why we went for a separate CT map approach rather than reusing the existing ones is that the original `CT_MAP_{TCP, ANY}{4,6}` maps are intended to really track connections, their state, stats, rev-nat information etc while this new `EGRESS_CT_MAP` map is supposed to track just which tuple is going to which gateway node, in order to pin it. So calling it a "connection table" might be a bit of a stretch, but given that also the Maglev paper calls it this way, I went for the same name.

At the same time I do agree that this information could be stored in the usual CT maps and that with this approach we are duplicating some data (specifically the `ipv4_tuple` key, across the 2 tables). Another approach would be to extend the existing maps with the `gateway_ip` field, but as you pointed out, that would cause some issues during upgrades/downgrades.

> At the same time, it's worth noting that by sizing the 'value' exactly to the size of the data you need for this PR, it also prevents this map from being resized in future without triggering the same sort of problems.

That's a good point, we could reserve some space in the value for future fields, but at the same time it feels bit arbitrary (i.e. how much space should we reserve?)

> Back in the day we had an open question "how big should the CT entry be to future-proof it?" and I think that question becomes relevant again here.

I just went for the same size we are using for the other CT maps since that size proved to work well in practice, assuming that `egress CT entries <= all CT entries` (while I think in practice it's safe to assume that `egress CT entries << all CT entries`

Reply:
> The reason why we went for a separate CT map approach rather than reusing the existing ones is that the original CT_MAP_{TCP, ANY}{4,6} maps are intended to really track connections, their state, stats, rev-nat information etc while this new EGRESS_CT_MAP map is supposed to track just which tuple is going to which gateway node, in order to pin it. So calling it a "connection table" might be a bit of a stretch, but given that also the Maglev paper calls it this way, I went for the same name.

For what it's worth, we do something similar already with entirely independent entries for the service stickiness, see `CT_SERVICE`. The `tuple->flags` contains values like `CT_INGRESS` / `CT_EGRESS` / `CT_SERVICE` to differentiate the types of info we are storing, and in the service case we are even overloading some of the ct_entry fields already. There may even be a middle ground there where we don't have to resize the ct_entry but we could store some other metadata about the flow for advanced features like this feature. It'd take one more entry per flow if you're using this feature, but that's not so bad, it's one less map to deal with.

> That's a good point, we could reserve some space in the value for future fields, but at the same time it feels bit arbitrary (i.e. how much space should we reserve?)

Yeah, this is the difficult part. One way I'm thinking about this is, could these entries be considered a more general "ct extension" where we can hang all sorts of information, and it just happens to be that the first feature that requires more metadata to be associated with a CT entry is this particular gateway stickiness info? You could imagine that over time we accumulate many similar maps like this, but ideally we could just say "alright depending on what you want to do, you may need to attach some additional info, and here's the mechanism to do this".

> > Back in the day we had an open question "how big should the CT entry be to future-proof it?" and I think that question becomes relevant again here.
> 
> I just went for the same size we are using for the other CT maps since that size proved to work well in practice, assuming that egress CT entries <= all CT entries (while I think in practice it's safe to assume that egress CT entries << all CT entries

Just to be clear, I was asking about the size of the ct_entry itself here (ie value), rather than the overall map size. This is just an extension of the same thoughts / musings about the size of the value above.

> Additional set of entries to handle garbage collection for, which can be expensive. In theory there should be tied fates with the corresponding entry in the main CT table right? But those tied fates must be implemented somehow, presumably through a tie-in to the main CT GC loop with an additional syscall per GC'd entry.

Just to bring this back up, even if we decide to make no further changes based on the above discussions, I wonder if we need something a bit more for this concern? Like today even with LRU enabled, we enable regular CT garbage collection to ensure we clean up the NAT map if any CT map entries are deleted. I think that the same logic would need to deal with this new map as well?

Reply:
OK, that's a good starting point. I'll play with the idea of storing the gateway related info in the usual CT maps.
I can introduce a new flag (to be used in the `ipv4_ct_tuple.flags` field), for example `TUPLE_F_EGRESS_GATEWAY` and then look for an existing field that can be overloaded. I'm not entirely sure about the implications of this approach for the rest of the datapath/if it's really doable, but worth giving it a try :+1: 

> Like today even with LRU enabled, we enable regular CT garbage collection to ensure we clean up the NAT map if any CT map entries are deleted. I think that the same logic would need to deal with this new map as well?

There's already some logic in place to delete these entries whenever the related policy gets deleted, but I'll double check if I'm missing anything